radix-engine/src/state_manager/staging.rs

```diff
@@ -0,0 +1,216 @@
+                self.remove_children(0);
+            }
+            return;
+        }
+
+        let node = self.nodes.remove(&id).unwrap();
+        if remove_children {
+            self.remove_children(id);
+        }
+
+        self.merge_to_parent_recurse(node.parent_id, true);
+
+        for (address, output_id) in node.spaces {
+            self.parent.put_space(&address, output_id);
+        }
+
+        for (address, output) in node.outputs {
+            self.parent.put_substate(&address, output);
+        }
+
+        if !remove_children {
+            self.set_root_parent(id);
+        }
+    }
+}
+
+pub struct ExecutionStore<'t, 's, S: ReadableSubstateStore + WriteableSubstateStore> {
+    stores: &'t mut StagedExecutionStores<'s, S>,
+    id: u64,
+}
+
+impl<'t, 's, S: ReadableSubstateStore + WriteableSubstateStore> ExecutionStore<'t, 's, S> {
+    fn get_substate_recurse(&self, address: &[u8], id: u64) -> Option<Output> {
+        if id == 0 {
+            return self.stores.parent.get_substate(address);
+        }
+
+        let node = self.stores.nodes.get(&id).unwrap();
+        if let Some(output) = node.outputs.get(address) {
+            // TODO: Remove encoding/decoding
+            let encoded_output = scrypto_encode(output);
+            return Some(scrypto_decode(&encoded_output).unwrap());
+        }
+
+        self.get_substate_recurse(address, node.parent_id)
+    }
+
+    fn get_space_recurse(&self, address: &[u8], id: u64) -> OutputId {
+        if id == 0 {
+            return self.stores.parent.get_space(address);
+        }
+
+        let node = self.stores.nodes.get(&id).unwrap();
+        if let Some(output_id) = node.spaces.get(address) {
+            return output_id.clone();
+        }
+
+        self.get_space_recurse(address, node.parent_id)
+    }
+}
+
+impl<'t, 's, S: ReadableSubstateStore + WriteableSubstateStore> ReadableSubstateStore
+    for ExecutionStore<'t, 's, S>
+{
+    fn get_substate(&self, address: &[u8]) -> Option<Output> {
+        self.get_substate_recurse(address, self.id)
+    }
+
+    fn get_space(&self, address: &[u8]) -> OutputId {
+        self.get_space_recurse(address, self.id)
+    }
+}
+
+impl<'t, 's, S: ReadableSubstateStore + WriteableSubstateStore> WriteableSubstateStore
+    for ExecutionStore<'t, 's, S>
+{
+    fn put_space(&mut self, address: &[u8], output_id: OutputId) {
+        if self.id == 0 {
+            self.stores.parent.put_space(address, output_id);
+        } else {
+            let node = self.stores.nodes.get_mut(&self.id).unwrap();
+            node.spaces.insert(address.to_vec(), output_id);
+        }
+    }
+
+    fn put_substate(&mut self, address: &[u8], output: Output) {
+        if self.id == 0 {
+            self.stores.parent.put_substate(address, output);
+        } else {
+            let node = self.stores.nodes.get_mut(&self.id).unwrap();
+            node.outputs.insert(address.to_vec(), output);
+        }
+    }
+}
+
```

Comment:
I'd also like to propose the following terms/names:

| Name                                             | Definition                                                                                                                                                                                                                |
|--------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `SubstateStore`                                  | (Not changed) the storage that physically stores substates, either in memory or in a database                                                                                                                             |
| `StagedSubstateStore`                            | A data structure that caches substate changes and points to a parent `SubstateStore` or `StagedSubstateStore` recursive substate reads/writes; it also implements the `ReadableSubstateStore` and `WriteableSubstateStore` traits. |
| `StagedSubstateStoreManager/Engine` | Manages the lifecyle of `StagedSubstateStore` and may also persisting them.

Reply:
These sound good.